### PR TITLE
suggest use of `kube explain <resource>` in kube get output

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:

``` release-note
release-note-none
```

This patch improves usability flow, making it easier for a user to
discover the command `kube explain <resource>` through `kube get` output.
##### After

```
$ kube get
You must specify the type of resource to get. Valid resource types include:
   * componentstatuses (aka 'cs')
   * configmaps (aka 'cm')
   * daemonsets (aka 'ds')
   * deployments (aka 'deploy')
   * events (aka 'ev')
   * endpoints (aka 'ep')
   * horizontalpodautoscalers (aka 'hpa')
   * ingress (aka 'ing')
   * jobs
   * limitranges (aka 'limits')
   * nodes (aka 'no')
   * namespaces (aka 'ns')
   * petsets (alpha feature, may be unstable)
   * pods (aka 'po')
   * persistentvolumes (aka 'pv')
   * persistentvolumeclaims (aka 'pvc')
   * quota
   * resourcequotas (aka 'quota')
   * replicasets (aka 'rs')
   * replicationcontrollers (aka 'rc')
   * secrets
   * serviceaccounts (aka 'sa')
   * services (aka 'svc')
error: Required resource not specified.
Use "kubectl explain <resource>" for a detailed description of that resource (e.g. kubectl explain pods).
See 'kubectl get -h' for help and examples.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31818)

### CODE REVIEW

**Comment by janetkuo** [comment]
ok to test

**Comment by janetkuo** [comment]
@kubernetes/kubectl 

Suggest not hardcoding "kubectl". You should be able to get the command name from `cmd.Parent().CommandPath()` (cobra command library)

**Comment by AdoHe** [comment]
@juanvallejo please do not use `kubectl` directly.

**Comment by juanvallejo** [comment]
@janetkuo @AdoHe Thanks for the feedback! I replaced `kubectl` with `cmd.Root().CommandPath()` https://github.com/kubernetes/kubernetes/pull/31818/files#diff-f23cc6c004fbd8a713f8465465142ab8R258

File: pkg/kubectl/cmd/get.go [hunk 1]
```diff
@@ -168,7 +180,8 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
 
 	if len(args) == 0 && cmdutil.IsFilenameEmpty(options.Filenames) {
 		fmt.Fprint(errOut, "You must specify the type of resource to get. ", valid_resources)
-		return cmdutil.UsageError(cmd, "Required resource not specified.")
+		usageString := fmt.Sprintf("Required resource not specified.\nUse \"%s explain <resource>\" for a detailed description of that resource (e.g. %[1]s explain pods).", options.FullCmdName)
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get.go]: You don't need to pass `FullCmdName` in. You can use the passed-in `cmd` and do the `cmd. cmds.Root().CommandPath()` here.

Comment (juanvallejo) [inline pkg/kubectl/cmd/get.go]: Done! PTAL https://github.com/kubernetes/kubernetes/pull/31818/commits/fea3756aeeaa1c6ca9b0611c2ef868807e506ecf

File: pkg/kubectl/cmd/get.go [hunk 2]
```diff
@@ -168,7 +168,14 @@ func RunGet(f *cmdutil.Factory, out io.Writer, errOut io.Writer, cmd *cobra.Comm
 
 	if len(args) == 0 && cmdutil.IsFilenameEmpty(options.Filenames) {
 		fmt.Fprint(errOut, "You must specify the type of resource to get. ", valid_resources)
-		return cmdutil.UsageError(cmd, "Required resource not specified.")
+
+		fullCmdName := cmd.Parent().CommandPath()
+		if fullCmdName == "" {
```
Comment (janetkuo) [inline pkg/kubectl/cmd/get.go]: If `fullCmdName == ""` we shouldn't assume that the user can use `kubectl explain`. 

We also need a check to see if `<parent_command> explain` exists, something like:
```go
func isExplainExist(cmd *cobra.Command) bool {
		for _, c := range cmd.Parent().Commands() {
			if c.Name() == "explain" {
				return true
			}
		}
		return false 
}
```

and we only print the suggestion when `len(fullCmdName)>0` (i.e. `fullCmdName != ""`) and the explain command is found:
```go
usageString := "Required resource not specified."
if len(fullCmdName)>0 && isExplainExist(cmd) {
	usageString = fmt.Sprintf("%s\nUse ...", usageString, fullCmdName)
}
```

**Comment by janetkuo** [comment]
One last suggestion. 
@juanvallejo would you open another PR (or a separate commit in this PR, whichever you prefer) to change the similar thing in `kubectl exec` to what this PR does? (i.e. remove passing "kubectl" / command path to `NewCmdExec` and do the similar check)

**Comment by juanvallejo** [comment]
@janetkuo Thanks for the feedback! 
> would you open another PR (or a separate commit in this PR, whichever you prefer) to change the similar thing in kubectl exec to what this PR does?

Sure, I'll add this in a new commit

File: pkg/kubectl/cmd/exec.go [hunk 3]
```diff
@@ -265,7 +267,11 @@ func (p *ExecOptions) Run() error {
 	containerName := p.ContainerName
 	if len(containerName) == 0 {
 		if len(pod.Spec.Containers) > 1 {
-			fmt.Fprintf(p.Err, "defaulting container name to %s, use '%s describe pod/%s' to see all container names\n", pod.Spec.Containers[0].Name, p.FullCmdName, p.PodName)
+			usageString := fmt.Sprintf("Defaulting container name to %s.", pod.Spec.Containers[0].Name)
+			if len(p.SuggestedCmdUsage) > 0 {
+				usageString = fmt.Sprintf("%s\n%s", usageString, p.SuggestedCmdUsage)
+			}
+			fmt.Fprintf(p.Err, "%s", usageString)
```
Comment (janetkuo) [inline pkg/kubectl/cmd/exec.go]: nit: add `\n` at the end

File: pkg/kubectl/cmd/util/helpers.go [hunk 4]
```diff
@@ -716,3 +716,15 @@ func ObjectListToVersionedObject(objects []runtime.Object, version unversioned.G
 	}
 	return converted, nil
 }
+
+// IsCommandExists receives a pointer to a cobra command and a target string.
+// Returns true if the target string is found in the list of sibling commands.
+func IsCommandExists(cmd *cobra.Command, targetCmdName string) bool {
```
Comment (janetkuo) [inline pkg/kubectl/cmd/util/helpers.go]: nit: on second thought we should name it `IsSibCommandExists` to make it more precise (update the name in the comment as well)

**Comment by janetkuo** [comment]
Only nits, lgtm otherwise. Feel free to squash the 1st, 2nd and 3rd commits into one.

File: pkg/kubectl/cmd/util/helpers.go [hunk 5]
```diff
@@ -716,3 +716,15 @@ func ObjectListToVersionedObject(objects []runtime.Object, version unversioned.G
 	}
 	return converted, nil
 }
+
+// IsSubCommandExists receives a pointer to a cobra command and a target string.
+// Returns true if the target string is found in the list of sibling commands.
+func IsSubCommandExists(cmd *cobra.Command, targetCmdName string) bool {
```
Comment (janetkuo) [inline pkg/kubectl/cmd/util/helpers.go]: typo: `IsSibCommandExists`, sib, not sub

Comment (janetkuo) [inline pkg/kubectl/cmd/util/helpers.go]: or should we use sibling instead of sib? In case people confuse sib with sub

Comment (juanvallejo) [inline pkg/kubectl/cmd/util/helpers.go]: Ah, I would prefer `sibling` instead, it would be a bit easier to understand

**Comment by janetkuo** [comment]
LGTM, thanks much!

**Comment by janetkuo** [comment]
@k8s-bot gke e2e test this

**Comment by janetkuo** [comment]
@k8s-bot gci gce e2e test this

**Pull request merged**